// vexip-ui/es/components/scroll/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, classProp, booleanProp, booleanNumberProp, eventProp } from "../../common/config/src/props.mjs";
const scrollProps = buildProps({
  scrollClass: classProp,
  mode: String,
  width: [Number, String],
  height: [Number, String],
  deltaX: Number,
  deltaY: Number,
  disabled: booleanProp,
  pointer: booleanProp,
  wheel: booleanProp,
  scrollX: Number,
  scrollY: Number,
  useXBar: booleanProp,
  useYBar: booleanProp,
  barFade: Number,
  barClass: classProp,
  autoplay: booleanNumberProp,
  playWaiting: Number,
  noBuffer: booleanProp,
  noTransition: booleanProp,
  onBeforeScroll: Function,
  useBarTrack: booleanProp,
  wrapperTag: String,
  onResize: eventProp(),
  onXEnabledChange: eventProp(),
  onYEnabledChange: eventProp(),
  onWheel: eventProp(),
  onScrollStart: eventProp(),
  onScroll: eventProp(),
  onScrollEnd: eventProp(),
  onBarScrollStart: eventProp(),
  onBarScroll: eventProp(),
  onBarScrollEnd: eventProp(),
  onReady: eventProp()
});
export { scrollProps };
