// vexip-ui/es/components/date-picker/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanProp, booleanStringProp, eventProp } from "../../common/config/src/props.mjs";
const datePickerProps = buildProps({
  size: sizeProp,
  state: stateProp,
  type: String,
  visible: booleanProp,
  placement: String,
  transfer: booleanStringProp,
  value: [Number, String, Date, Array],
  format: String,
  filler: String,
  noFiller: booleanProp,
  clearable: booleanProp,
  noAction: booleanProp,
  labels: Object,
  dateSeparator: String,
  timeSeparator: String,
  shortcuts: Array,
  disabledDate: Function,
  steps: Array,
  ctrlSteps: Array,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  noSuffix: booleanProp,
  disabled: booleanProp,
  transitionName: String,
  confirmText: String,
  cancelText: String,
  today: [Number, String, Date],
  isRange: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  min: [Number, String, Date],
  max: [Number, String, Date],
  onInput: eventProp(),
  onPlus: eventProp(),
  onMinus: eventProp(),
  onEnter: eventProp(),
  onCancel: eventProp(),
  onChange: eventProp(),
  onClear: eventProp(),
  onShortcut: eventProp(),
  onToggle: eventProp(),
  onFocus: eventProp(),
  onBlur: eventProp(),
  onChangeCol: eventProp(),
  onClickOutside: eventProp()
});
const timePickerProps = buildProps({
  size: sizeProp,
  state: stateProp,
  visible: booleanProp,
  placement: String,
  transfer: booleanStringProp,
  format: String,
  separator: String,
  value: [String, Array],
  filler: String,
  noFiller: booleanProp,
  clearable: booleanProp,
  noAction: booleanProp,
  noArrow: booleanProp,
  pointer: booleanProp,
  candidate: Number,
  steps: Array,
  labels: Object,
  shortcuts: Array,
  isRange: booleanProp,
  disabled: booleanProp,
  transitionName: String,
  confirmText: String,
  cancelText: String,
  ctrlSteps: Array,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  noSuffix: booleanProp,
  exchange: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  min: String,
  max: String,
  onInput: eventProp(),
  onPlus: eventProp(),
  onMinus: eventProp(),
  onEnter: eventProp(),
  onCancel: eventProp(),
  onChange: eventProp(),
  onClear: eventProp(),
  onShortcut: eventProp(),
  onToggle: eventProp(),
  onFocus: eventProp(),
  onBlur: eventProp(),
  onChangeCol: eventProp(),
  onClickOutside: eventProp()
});
export { datePickerProps, timePickerProps };
