// vexip-ui/es/components/layout/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanProp, booleanStringProp, eventProp } from "../../common/config/src/props.mjs";
const layoutProps = buildProps({
  noAside: booleanProp,
  footer: booleanProp,
  tag: String,
  menus: Object,
  menuProps: Object,
  logo: String,
  signName: String,
  config: Array,
  user: Object,
  actions: Array,
  reduced: booleanProp,
  avatarCircle: booleanProp,
  signType: String,
  headerFixed: booleanStringProp,
  asideFixed: booleanStringProp,
  copyright: String,
  links: Array,
  colors: Array,
  color: String,
  miniHeaderSign: booleanStringProp,
  verticalLinks: booleanStringProp,
  onReducedChange: eventProp(),
  onSignClick: eventProp(),
  onMenuSelect: eventProp(),
  onUserAction: eventProp(),
  onNavChange: eventProp(),
  onColorChange: eventProp()
});
const layoutAsideProps = buildProps({
  tag: String,
  expanded: booleanProp,
  reduced: booleanProp,
  menus: Array,
  menuProps: Object,
  logo: String,
  signName: String,
  fixed: booleanStringProp,
  onReducedChange: eventProp(),
  onExpandedChange: eventProp(),
  onSignClick: eventProp(),
  onMenuSelect: eventProp()
});
const layoutHeaderProps = buildProps({
  tag: String,
  logo: String,
  signName: String,
  user: Object,
  userDropped: booleanProp,
  avatarCircle: booleanProp,
  config: Array,
  actions: Array,
  signType: String,
  colors: Array,
  color: String,
  menus: Object,
  menuProps: Object,
  onNavChange: eventProp(),
  onColorChange: eventProp(),
  onUserAction: eventProp(),
  onSignClick: eventProp(),
  onDropChange: eventProp(),
  onReducedChange: eventProp(),
  onMenuSelect: eventProp()
});
const layoutFooterProps = buildProps({
  tag: String,
  copyright: String,
  links: Array,
  verticalLinks: booleanStringProp
});
export { layoutAsideProps, layoutFooterProps, layoutHeaderProps, layoutProps };
