// vexip-ui/es/components/modal/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanStringProp, booleanProp, classProp, styleProp, eventProp } from "../../common/config/src/props.mjs";
const positionType = [Number, String];
const modalProps = buildProps({
  transfer: booleanStringProp,
  active: booleanProp,
  width: positionType,
  height: positionType,
  top: positionType,
  left: positionType,
  right: positionType,
  bottom: positionType,
  title: String,
  closable: booleanProp,
  inner: booleanProp,
  maskClose: booleanProp,
  modalClass: classProp,
  modalStyle: styleProp,
  noFooter: booleanProp,
  hideMask: booleanProp,
  draggable: booleanProp,
  resizable: booleanProp,
  onBeforeClose: Function,
  loading: booleanProp,
  minWidth: Number,
  minHeight: Number,
  transitionName: String,
  confirmText: String,
  cancelText: String,
  autoRemove: booleanProp,
  onToggle: eventProp(),
  onConfirm: eventProp(),
  onCancel: eventProp(),
  onClose: eventProp(),
  onShow: eventProp(),
  onHide: eventProp(),
  onDragStart: eventProp(),
  onDragMove: eventProp(),
  onDragEnd: eventProp(),
  onResizeStart: eventProp(),
  onResizeMove: eventProp(),
  onResizeEnd: eventProp()
});
export { modalProps };
