// vexip-ui/es/components/typography/h.mjs
import { defineComponent, createVNode, mergeProps } from "vue";
import "../../common/config/src/index.mjs";
import Title from "./title.mjs";
import { hProps } from "./props.mjs";
import { useProps } from "../../common/config/src/props.mjs";
function createHComponent(level) {
  return defineComponent({
    name: `H${level}`,
    props: hProps,
    emits: [],
    setup(_props, {
      slots
    }) {
      const props = useProps(`h${level}`, _props, {
        type: "default",
        top: false,
        marker: false,
        aligned: false,
        thin: false
      });
      return () => createVNode(Title, mergeProps(props, {
        "level": level
      }), {
        default: () => {
          var _a;
          return [(_a = slots.default) == null ? void 0 : _a.call(slots)];
        }
      });
    }
  });
}
const H1 = createHComponent(1);
const H2 = createHComponent(2);
const H3 = createHComponent(3);
const H4 = createHComponent(4);
const H5 = createHComponent(5);
const H6 = createHComponent(6);
export { H1, H2, H3, H4, H5, H6 };
