// vexip-ui/es/components/typography/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanProp, omitProps } from "../../common/config/src/props.mjs";
const textProps = buildProps({
  type: String,
  tag: String,
  delete: booleanProp,
  strong: booleanProp,
  italic: booleanProp,
  underline: booleanProp,
  code: booleanProp,
  mark: booleanProp,
  disabled: booleanProp,
  keyboard: booleanProp,
  thin: booleanProp,
  reversed: booleanProp
});
const pProps = omitProps(textProps, ["tag", "code"]);
const strongProps = omitProps(textProps, ["tag", "strong", "code"]);
const titleProps = buildProps({
  type: String,
  level: Number,
  top: booleanProp,
  marker: booleanProp,
  aligned: booleanProp,
  thin: booleanProp,
  markerType: String
});
const hProps = omitProps(titleProps, ["level"]);
const h1Props = hProps;
const h2Props = hProps;
const h3Props = hProps;
const h4Props = hProps;
const h5Props = hProps;
const h6Props = hProps;
const blockquoteProps = buildProps({
  type: String
});
const olProps = buildProps({
  type: String
});
const ulProps = buildProps({
  listStyle: String
});
export { blockquoteProps, h1Props, h2Props, h3Props, h4Props, h5Props, h6Props, hProps, olProps, pProps, strongProps, textProps, titleProps, ulProps };
