// vexip-ui/es/components/table/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanProp, eventProp, classProp, styleProp, sizeProp } from "../../common/config/src/props.mjs";
const tableProps = buildProps({
  columns: Array,
  data: Array,
  dataKey: String,
  width: [Number, String],
  height: Number,
  rowClass: [String, Object, Array, Function],
  rowStyle: [String, Object, Array, Function],
  rowAttrs: [Object, Function],
  stripe: booleanProp,
  border: booleanProp,
  highlight: booleanProp,
  useYBar: booleanProp,
  barFade: Number,
  scrollDeltaY: Number,
  rowDraggable: booleanProp,
  rowHeight: Number,
  rowMinHeight: Number,
  virtual: booleanProp,
  bufferCount: Number,
  scrollClass: Object,
  expandRenderer: Function,
  currentPage: Number,
  pageSize: Number,
  transparent: booleanProp,
  emptyText: String,
  tooltipTheme: String,
  tooltipWidth: [Number, String],
  singleSorter: booleanProp,
  singleFilter: booleanProp,
  cellClass: [String, Object, Array, Function],
  cellStyle: [String, Object, Array, Function],
  cellAttrs: [Object, Function],
  headClass: [String, Object, Array, Function],
  headStyle: [String, Object, Array, Function],
  headAttrs: [Object, Function],
  onBodyScroll: eventProp(),
  onRowEnter: eventProp(),
  onRowLeave: eventProp(),
  onRowClick: eventProp(),
  onRowDblclick: eventProp(),
  onRowContextmenu: eventProp(),
  onRowCheck: eventProp(),
  onRowCheckAll: eventProp(),
  onRowExpand: eventProp(),
  onRowDragStart: eventProp(),
  onRowDragOver: eventProp(),
  onRowDrop: eventProp(),
  onRowDragEnd: eventProp(),
  onRowFilter: eventProp(),
  onRowSort: eventProp(),
  onCellEnter: eventProp(),
  onCellLeave: eventProp(),
  onCellClick: eventProp(),
  onCellDblclick: eventProp(),
  onCellContextmenu: eventProp(),
  onHeadEnter: eventProp(),
  onHeadLeave: eventProp(),
  onHeadClick: eventProp(),
  onHeadDblclick: eventProp(),
  onHeadContextmenu: eventProp()
});
const tableColumnProps = buildProps({
  idKey: [Number, String],
  name: String,
  accessor: Function,
  fixed: {
    type: [Boolean, String],
    default: null
  },
  className: classProp,
  style: styleProp,
  attrs: Object,
  type: String,
  width: Number,
  filter: Object,
  sorter: {
    type: [Boolean, Object],
    default: null
  },
  renderer: Function,
  headRenderer: Function,
  order: Number,
  noEllipsis: booleanProp,
  checkboxSize: sizeProp,
  disableRow: Function,
  truthIndex: booleanProp,
  orderLabel: Function,
  metaData: Object
});
export { tableColumnProps, tableProps };
