// vexip-ui/es/components/select/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanProp, classProp, booleanStringProp, eventProp } from "../../common/config/src/props.mjs";
const selectProps = buildProps({
  size: sizeProp,
  state: stateProp,
  visible: booleanProp,
  options: Array,
  disabled: booleanProp,
  transitionName: String,
  outsideClose: booleanProp,
  placeholder: String,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  noSuffix: booleanProp,
  value: [String, Number, Array],
  multiple: booleanProp,
  clearable: booleanProp,
  maxListHeight: Number,
  listClass: classProp,
  placement: String,
  transfer: booleanStringProp,
  optionCheck: booleanProp,
  emptyText: String,
  staticSuffix: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  keyConfig: Object,
  filter: {
    type: [Boolean, Function],
    default: null
  },
  ignoreCase: booleanProp,
  creatable: booleanProp,
  transparent: booleanProp,
  onFocus: eventProp(),
  onBlur: eventProp(),
  onToggle: eventProp(),
  onSelect: eventProp(),
  onCancel: eventProp(),
  onChange: eventProp(),
  onClickOutside: eventProp(),
  onOutsideClose: eventProp(),
  onClear: eventProp()
});
export { selectProps };
