// vexip-ui/es/components/cascader/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanProp, booleanStringProp, eventProp } from "../../common/config/src/props.mjs";
const cascaderProps = buildProps({
  size: sizeProp,
  state: stateProp,
  value: Array,
  visible: booleanProp,
  options: Array,
  placeholder: String,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  noCascaded: booleanProp,
  multiple: booleanProp,
  disabled: booleanProp,
  clearable: booleanProp,
  placement: String,
  transfer: booleanStringProp,
  staticSuffix: booleanProp,
  noSuffix: booleanProp,
  transitionName: String,
  outsideClose: booleanProp,
  keyConfig: Object,
  separator: String,
  hoverTrigger: booleanProp,
  maxTagCount: Number,
  briefLabel: booleanProp,
  noRestTip: booleanProp,
  onAsyncLoad: Function,
  mergeTags: booleanProp,
  tagType: String,
  emptyText: String,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  transparent: booleanProp,
  onToggle: eventProp(),
  onSelect: eventProp(),
  onCancel: eventProp(),
  onChange: eventProp(),
  onClickOutside: eventProp(),
  onOutsideClose: eventProp(),
  onClear: eventProp()
});
export { cascaderProps };
