// vexip-ui/es/components/index.mjs
import "./alert/index.mjs";
import "./anchor/index.mjs";
import "./anchor-link/index.mjs";
import "./auto-complete/index.mjs";
import "./avatar/index.mjs";
import "./avatar-group/index.mjs";
import "./badge/index.mjs";
import "./breadcrumb/index.mjs";
import "./breadcrumb-item/index.mjs";
import "./bubble/index.mjs";
import "./button/index.mjs";
import "./button-group/index.mjs";
import "./calendar/index.mjs";
import "./calendar-panel/index.mjs";
import "./card/index.mjs";
import "./carousel/index.mjs";
import "./carousel-item/index.mjs";
import "./cascader/index.mjs";
import "./cell/index.mjs";
import "./checkbox/index.mjs";
import "./checkbox-group/index.mjs";
import "./collapse/index.mjs";
import "./collapse-panel/index.mjs";
import "./collapse-transition/index.mjs";
import "./color-picker/index.mjs";
import "./column/index.mjs";
import "./config-provider/index.mjs";
import { Confirm } from "./confirm/index.mjs";
export { Confirm, ConfirmManager } from "./confirm/index.mjs";
import { Contextmenu } from "./contextmenu/index.mjs";
export { Contextmenu, ContextmenuManager } from "./contextmenu/index.mjs";
import "./date-picker/index.mjs";
import "./divider/index.mjs";
import "./drawer/index.mjs";
import "./dropdown/index.mjs";
import "./dropdown-item/index.mjs";
import "./dropdown-list/index.mjs";
import "./ellipsis/index.mjs";
import "./form/index.mjs";
import "./form-item/index.mjs";
import "./form-reset/index.mjs";
import "./form-submit/index.mjs";
import "./grid/index.mjs";
import "./highlight/index.mjs";
import "./icon/index.mjs";
import "./input/index.mjs";
import "./layout/index.mjs";
import "./layout-aside/index.mjs";
import "./layout-footer/index.mjs";
import "./layout-header/index.mjs";
import "./linker/index.mjs";
import { Loading } from "./loading/index.mjs";
export { Loading, LoadingManager } from "./loading/index.mjs";
import "./masker/index.mjs";
import "./menu/index.mjs";
import "./menu-group/index.mjs";
import "./menu-item/index.mjs";
import { Message } from "./message/index.mjs";
export { Message, MessageManager } from "./message/index.mjs";
import "./modal/index.mjs";
import "./native-scroll/index.mjs";
import { Notice } from "./notice/index.mjs";
export { Notice, NoticeManager } from "./notice/index.mjs";
import "./number-input/index.mjs";
import "./option/index.mjs";
import "./option-group/index.mjs";
import "./overflow/index.mjs";
import "./pagination/index.mjs";
import "./popup/index.mjs";
import "./portal/index.mjs";
import "./progress/index.mjs";
import "./radio/index.mjs";
import "./radio-group/index.mjs";
import "./renderer/index.mjs";
import "./resize-observer/index.mjs";
import "./row/index.mjs";
import "./scroll/index.mjs";
import "./scrollbar/index.mjs";
import "./select/index.mjs";
import "./skeleton/index.mjs";
import "./skeleton-group/index.mjs";
import "./slider/index.mjs";
import "./space/index.mjs";
import "./spin/index.mjs";
import "./split/index.mjs";
import "./switch/index.mjs";
import "./tab-nav/index.mjs";
import "./tab-nav-item/index.mjs";
import "./tab-panel/index.mjs";
import "./table/index.mjs";
import "./table-column/index.mjs";
import "./tabs/index.mjs";
import "./tag/index.mjs";
import "./textarea/index.mjs";
import "./time-ago/index.mjs";
import "./time-picker/index.mjs";
import "./timeline/index.mjs";
import "./timeline-item/index.mjs";
import { Toast } from "./toast/index.mjs";
export { Toast, ToastManager } from "./toast/index.mjs";
import "./tooltip/index.mjs";
import "./transfer/index.mjs";
import "./tree/index.mjs";
import "./upload/index.mjs";
import "./upload-file/index.mjs";
import "./upload-list/index.mjs";
import "./viewer/index.mjs";
import "./virtual-list/index.mjs";
import "./wheel/index.mjs";
import "./typography/index.mjs";
import { buildInstall } from "./create.mjs";
export { buildInstall } from "./create.mjs";
import { install as install$1 } from "../directives/index.mjs";
export { version } from "./version.mjs";
export { vLoading } from "../directives/loading/index.mjs";
import Alert from "./alert/alert.mjs";
export { default as Alert } from "./alert/alert.mjs";
import Anchor from "./anchor/anchor.mjs";
export { default as Anchor } from "./anchor/anchor.mjs";
import AnchorLink from "./anchor/anchor-link.mjs";
export { default as AnchorLink } from "./anchor/anchor-link.mjs";
import AutoComplete from "./auto-complete/auto-complete.mjs";
export { default as AutoComplete } from "./auto-complete/auto-complete.mjs";
import Avatar from "./avatar/avatar.mjs";
export { default as Avatar } from "./avatar/avatar.mjs";
import AvatarGroup from "./avatar/avatar-group.mjs";
export { default as AvatarGroup } from "./avatar/avatar-group.mjs";
import Badge from "./badge/badge.mjs";
export { default as Badge } from "./badge/badge.mjs";
import Breadcrumb from "./breadcrumb/breadcrumb.mjs";
export { default as Breadcrumb } from "./breadcrumb/breadcrumb.mjs";
import BreadcrumbItem from "./breadcrumb/breadcrumb-item.mjs";
export { default as BreadcrumbItem } from "./breadcrumb/breadcrumb-item.mjs";
import Bubble from "./bubble/bubble.mjs";
export { default as Bubble } from "./bubble/bubble.mjs";
import Button from "./button/button.mjs";
export { default as Button } from "./button/button.mjs";
import ButtonGroup from "./button/button-group.mjs";
export { default as ButtonGroup } from "./button/button-group.mjs";
import Calendar from "./calendar/calendar.mjs";
export { default as Calendar } from "./calendar/calendar.mjs";
import CalendarPanel from "./calendar/calendar-panel.mjs";
export { default as CalendarPanel } from "./calendar/calendar-panel.mjs";
import Card from "./card/card.mjs";
export { default as Card } from "./card/card.mjs";
import Carousel from "./carousel/carousel.mjs";
export { default as Carousel } from "./carousel/carousel.mjs";
import CarouselItem from "./carousel/carousel-item.mjs";
export { default as CarouselItem } from "./carousel/carousel-item.mjs";
import Cascader from "./cascader/cascader.mjs";
export { default as Cascader } from "./cascader/cascader.mjs";
import Cell from "./grid/cell.mjs";
export { default as Cell } from "./grid/cell.mjs";
import Checkbox from "./checkbox/checkbox.mjs";
export { default as Checkbox } from "./checkbox/checkbox.mjs";
import CheckboxGroup from "./checkbox/checkbox-group.mjs";
export { default as CheckboxGroup } from "./checkbox/checkbox-group.mjs";
import Collapse from "./collapse/collapse.mjs";
export { default as Collapse } from "./collapse/collapse.mjs";
import CollapsePanel from "./collapse/collapse-panel.mjs";
export { default as CollapsePanel } from "./collapse/collapse-panel.mjs";
import CollapseTransition from "./collapse/collapse-transition.mjs";
export { default as CollapseTransition } from "./collapse/collapse-transition.mjs";
import ColorPicker from "./color-picker/color-picker.mjs";
export { default as ColorPicker } from "./color-picker/color-picker.mjs";
import Column from "./row/column.mjs";
export { default as Column } from "./row/column.mjs";
import ConfigProvider from "./config-provider/config-provider.mjs";
export { default as ConfigProvider } from "./config-provider/config-provider.mjs";
import DatePicker from "./date-picker/date-picker.mjs";
export { default as DatePicker } from "./date-picker/date-picker.mjs";
import Divider from "./divider/divider.mjs";
export { default as Divider } from "./divider/divider.mjs";
import Drawer from "./drawer/drawer.mjs";
export { default as Drawer } from "./drawer/drawer.mjs";
import Dropdown from "./dropdown/dropdown.mjs";
export { default as Dropdown } from "./dropdown/dropdown.mjs";
import DropdownItem from "./dropdown/dropdown-item.mjs";
export { default as DropdownItem } from "./dropdown/dropdown-item.mjs";
import DropdownList from "./dropdown/dropdown-list.mjs";
export { default as DropdownList } from "./dropdown/dropdown-list.mjs";
import Ellipsis from "./ellipsis/ellipsis.mjs";
export { default as Ellipsis } from "./ellipsis/ellipsis.mjs";
import Form from "./form/form.mjs";
export { default as Form } from "./form/form.mjs";
import FormItem from "./form/form-item.mjs";
export { default as FormItem } from "./form/form-item.mjs";
import FormReset from "./form/form-reset.mjs";
export { default as FormReset } from "./form/form-reset.mjs";
import FormSubmit from "./form/form-submit.mjs";
export { default as FormSubmit } from "./form/form-submit.mjs";
import Grid from "./grid/grid.mjs";
export { default as Grid } from "./grid/grid.mjs";
import Highlight from "./highlight/highlight.mjs";
export { default as Highlight } from "./highlight/highlight.mjs";
import Icon from "./icon/icon.mjs";
export { default as Icon } from "./icon/icon.mjs";
import Input from "./input/input.mjs";
export { default as Input } from "./input/input.mjs";
import Layout from "./layout/layout.mjs";
export { default as Layout } from "./layout/layout.mjs";
import LayoutAside from "./layout/layout-aside.mjs";
export { default as LayoutAside } from "./layout/layout-aside.mjs";
import LayoutFooter from "./layout/layout-footer.mjs";
export { default as LayoutFooter } from "./layout/layout-footer.mjs";
import LayoutHeader from "./layout/layout-header.mjs";
export { default as LayoutHeader } from "./layout/layout-header.mjs";
import Linker from "./linker/linker.mjs";
export { default as Linker } from "./linker/linker.mjs";
import Masker from "./masker/masker.mjs";
export { default as Masker } from "./masker/masker.mjs";
import Menu from "./menu/menu.mjs";
export { default as Menu } from "./menu/menu.mjs";
import MenuGroup from "./menu/menu-group.mjs";
export { default as MenuGroup } from "./menu/menu-group.mjs";
import MenuItem from "./menu/menu-item.mjs";
export { default as MenuItem } from "./menu/menu-item.mjs";
import Modal from "./modal/modal.mjs";
export { default as Modal } from "./modal/modal.mjs";
import NativeScroll from "./native-scroll/native-scroll.mjs";
export { default as NativeScroll } from "./native-scroll/native-scroll.mjs";
import NumberInput from "./number-input/number-input.mjs";
export { default as NumberInput } from "./number-input/number-input.mjs";
import Option from "./option/option.mjs";
export { default as Option } from "./option/option.mjs";
import OptionGroup from "./option/option-group.mjs";
export { default as OptionGroup } from "./option/option-group.mjs";
import Overflow from "./overflow/overflow.mjs";
export { default as Overflow } from "./overflow/overflow.mjs";
import Pagination from "./pagination/pagination.mjs";
export { default as Pagination } from "./pagination/pagination.mjs";
import Popup from "./popup/popup.mjs";
export { default as Popup } from "./popup/popup.mjs";
import Portal from "./portal/portal.mjs";
export { default as Portal } from "./portal/portal.mjs";
import Progress from "./progress/progress.mjs";
export { default as Progress } from "./progress/progress.mjs";
import Radio from "./radio/radio.mjs";
export { default as Radio } from "./radio/radio.mjs";
import RadioGroup from "./radio/radio-group.mjs";
export { default as RadioGroup } from "./radio/radio-group.mjs";
import Renderer from "./renderer/renderer.mjs";
export { default as Renderer } from "./renderer/renderer.mjs";
import ResizeObserver from "./resize-observer/resize-observer.mjs";
export { default as ResizeObserver } from "./resize-observer/resize-observer.mjs";
import Row from "./row/row.mjs";
export { default as Row } from "./row/row.mjs";
import Scroll from "./scroll/scroll.mjs";
export { default as Scroll } from "./scroll/scroll.mjs";
import Scrollbar from "./scrollbar/scrollbar.mjs";
export { default as Scrollbar } from "./scrollbar/scrollbar.mjs";
import Select from "./select/select.mjs";
export { default as Select } from "./select/select.mjs";
import Skeleton from "./skeleton/skeleton.mjs";
export { default as Skeleton } from "./skeleton/skeleton.mjs";
import SkeletonGroup from "./skeleton/skeleton-group.mjs";
export { default as SkeletonGroup } from "./skeleton/skeleton-group.mjs";
import Slider from "./slider/slider.mjs";
export { default as Slider } from "./slider/slider.mjs";
import Space from "./space/space.mjs";
export { default as Space } from "./space/space.mjs";
import Spin from "./spin/spin.mjs";
export { default as Spin } from "./spin/spin.mjs";
import Split from "./split/split.mjs";
export { default as Split } from "./split/split.mjs";
import Switch from "./switch/switch.mjs";
export { default as Switch } from "./switch/switch.mjs";
import TabNav from "./tab-nav/tab-nav.mjs";
export { default as TabNav } from "./tab-nav/tab-nav.mjs";
import TabNavItem from "./tab-nav/tab-nav-item.mjs";
export { default as TabNavItem } from "./tab-nav/tab-nav-item.mjs";
import TabPanel from "./tabs/tab-panel.mjs";
export { default as TabPanel } from "./tabs/tab-panel.mjs";
import Table from "./table/table.mjs";
export { default as Table } from "./table/table.mjs";
import TableColumn from "./table/table-column.mjs";
export { default as TableColumn } from "./table/table-column.mjs";
import Tabs from "./tabs/tabs.mjs";
export { default as Tabs } from "./tabs/tabs.mjs";
import Tag from "./tag/tag.mjs";
export { default as Tag } from "./tag/tag.mjs";
import Textarea from "./textarea/textarea.mjs";
export { default as Textarea } from "./textarea/textarea.mjs";
import TimeAgo from "./time-ago/time-ago.mjs";
export { default as TimeAgo } from "./time-ago/time-ago.mjs";
import TimePicker from "./date-picker/time-picker.mjs";
export { default as TimePicker } from "./date-picker/time-picker.mjs";
import Timeline from "./timeline/timeline.mjs";
export { default as Timeline } from "./timeline/timeline.mjs";
import TimelineItem from "./timeline/timeline-item.mjs";
export { default as TimelineItem } from "./timeline/timeline-item.mjs";
import Tooltip from "./tooltip/tooltip.mjs";
export { default as Tooltip } from "./tooltip/tooltip.mjs";
import Transfer from "./transfer/transfer.mjs";
export { default as Transfer } from "./transfer/transfer.mjs";
import Tree from "./tree/tree.mjs";
export { default as Tree } from "./tree/tree.mjs";
import Upload from "./upload/upload.mjs";
export { default as Upload } from "./upload/upload.mjs";
import UploadFile from "./upload/upload-file.mjs";
export { default as UploadFile } from "./upload/upload-file.mjs";
import UploadList from "./upload/upload-list.mjs";
export { default as UploadList } from "./upload/upload-list.mjs";
import Viewer from "./viewer/viewer.mjs";
export { default as Viewer } from "./viewer/viewer.mjs";
import VirtualList from "./virtual-list/virtual-list.mjs";
export { default as VirtualList } from "./virtual-list/virtual-list.mjs";
import Wheel from "./wheel/wheel.mjs";
export { default as Wheel } from "./wheel/wheel.mjs";
import Title from "./typography/title.mjs";
export { default as Title } from "./typography/title.mjs";
import Text from "./typography/text.mjs";
export { default as Text } from "./typography/text.mjs";
import Blockquote from "./typography/blockquote.mjs";
export { default as Blockquote } from "./typography/blockquote.mjs";
import OL from "./typography/ol.mjs";
export { default as OL } from "./typography/ol.mjs";
import UL from "./typography/ul.mjs";
export { default as UL } from "./typography/ul.mjs";
import { H1, H2, H3, H4, H5, H6 } from "./typography/h.mjs";
export { H1, H2, H3, H4, H5, H6 } from "./typography/h.mjs";
import P from "./typography/p.mjs";
export { default as P } from "./typography/p.mjs";
import Strong from "./typography/strong.mjs";
export { default as Strong } from "./typography/strong.mjs";
export { getValueByPath, setValueByPath, useFieldStore } from "./form/helper.mjs";
export { validate, validateBoolean, validateDate, validateEmail, validateEnumeration, validateFloat, validateInt, validateNumber, validateRequirement, validateString, validateType, validateUrl } from "./form/validator.mjs";
export { currentBreakPoint } from "./grid/helpler.mjs";
export { animateScrollTo } from "./native-scroll/helper.mjs";
export { defineColumn, defineColumns, defineFilter, defineSorter } from "./table/helper.mjs";
export { iconMaps as uploadFileIconMap } from "./upload/file-icon.mjs";
const components = [
  Alert,
  Anchor,
  AnchorLink,
  AutoComplete,
  Avatar,
  AvatarGroup,
  Badge,
  Breadcrumb,
  BreadcrumbItem,
  Bubble,
  Button,
  ButtonGroup,
  Calendar,
  CalendarPanel,
  Card,
  Carousel,
  CarouselItem,
  Cascader,
  Cell,
  Checkbox,
  CheckboxGroup,
  Collapse,
  CollapsePanel,
  CollapseTransition,
  ColorPicker,
  Column,
  ConfigProvider,
  DatePicker,
  Divider,
  Drawer,
  Dropdown,
  DropdownItem,
  DropdownList,
  Ellipsis,
  Form,
  FormItem,
  FormReset,
  FormSubmit,
  Grid,
  Highlight,
  Icon,
  Input,
  Layout,
  LayoutAside,
  LayoutFooter,
  LayoutHeader,
  Linker,
  Masker,
  Menu,
  MenuGroup,
  MenuItem,
  Modal,
  NativeScroll,
  NumberInput,
  Option,
  OptionGroup,
  Overflow,
  Pagination,
  Popup,
  Portal,
  Progress,
  Radio,
  RadioGroup,
  Renderer,
  ResizeObserver,
  Row,
  Scroll,
  Scrollbar,
  Select,
  Skeleton,
  SkeletonGroup,
  Slider,
  Space,
  Spin,
  Split,
  Switch,
  TabNav,
  TabNavItem,
  TabPanel,
  Table,
  TableColumn,
  Tabs,
  Tag,
  Textarea,
  TimeAgo,
  TimePicker,
  Timeline,
  TimelineItem,
  Tooltip,
  Transfer,
  Tree,
  Upload,
  UploadFile,
  UploadList,
  Viewer,
  VirtualList,
  Wheel,
  Confirm,
  Contextmenu,
  Loading,
  Message,
  Notice,
  Toast,
  Title,
  Text,
  Blockquote,
  OL,
  UL,
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
  P,
  Strong,
  install$1
];
const install = buildInstall(components);
export { install };
