// vexip-ui/es/components/carousel/carousel-item.mjs
import { defineComponent, inject, reactive, computed, onBeforeUnmount, openBlock, createElementBlock, normalizeClass, normalizeStyle, renderSlot } from "vue";
import "../../common/config/src/index.mjs";
import { CAROUSEL_STATE } from "./symbol.mjs";
import _export_sfc from "../../_virtual/plugin-vue_export-helper.mjs";
import { useNameHelper } from "../../common/config/src/namespace.mjs";
const _sfc_main = defineComponent({
  name: "CarouselItem",
  setup() {
    const carouselState = inject(CAROUSEL_STATE, null);
    const nh = useNameHelper("carousel");
    const state = reactive({
      label: -1,
      width: 0,
      height: 0,
      offset: 0
    });
    const active = computed(() => {
      var _a;
      return (_a = carouselState == null ? void 0 : carouselState.isItemActive(state.label)) != null ? _a : false;
    });
    const className = computed(() => {
      return {
        [nh.be("item")]: true,
        [nh.bem("item", "active")]: active.value
      };
    });
    const style = computed(() => {
      const { width, height, offset } = state;
      return {
        width: width ? `${width}px` : void 0,
        height: height ? `${height}px` : void 0,
        transform: offset ? `translate${(carouselState == null ? void 0 : carouselState.vertical) ? "Y" : "X"}(${offset}px)` : void 0
      };
    });
    if (carouselState) {
      carouselState.increaseItem(state);
      onBeforeUnmount(() => {
        carouselState.decreaseItem(state);
      });
    }
    function handleClick() {
      carouselState == null ? void 0 : carouselState.handleSelect(state.label);
    }
    return {
      active,
      className,
      style,
      handleClick
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.className),
    role: "listitem",
    style: normalizeStyle(_ctx.style),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    renderSlot(_ctx.$slots, "default", { active: _ctx.active })
  ], 6);
}
var CarouselItem = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { CarouselItem as default };
