// vexip-ui/es/components/option/option-group.mjs
import { defineComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, renderSlot, createTextVNode, toDisplayString } from "vue";
import "../../common/config/src/index.mjs";
import _export_sfc from "../../_virtual/plugin-vue_export-helper.mjs";
import { useNameHelper } from "../../common/config/src/namespace.mjs";
const _sfc_main = defineComponent({
  name: "OptionGroup",
  props: {
    label: {
      type: String,
      default: ""
    },
    divided: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      nh: useNameHelper("option-group")
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("li", {
    class: normalizeClass({
      [_ctx.nh.b()]: true,
      [_ctx.nh.ns("option-vars")]: true,
      [_ctx.nh.bm("divided")]: _ctx.divided
    })
  }, [
    createElementVNode("div", {
      class: normalizeClass(_ctx.nh.be("label"))
    }, [
      renderSlot(_ctx.$slots, "label", {}, () => [
        createTextVNode(toDisplayString(_ctx.label), 1)
      ])
    ], 2),
    createElementVNode("ul", {
      class: normalizeClass(_ctx.nh.be("list"))
    }, [
      renderSlot(_ctx.$slots, "default")
    ], 2)
  ], 2);
}
var OptionGroup = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { OptionGroup as default };
