// vexip-ui/es/components/native-scroll/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, classProp, styleProp, booleanProp, booleanNumberProp, eventProp } from "../../common/config/src/props.mjs";
const nativeScrollProps = buildProps({
  scrollClass: classProp,
  scrollStyle: styleProp,
  mode: String,
  width: [Number, String],
  height: [Number, String],
  disabled: booleanProp,
  pointer: booleanProp,
  scrollX: Number,
  scrollY: Number,
  useXBar: booleanProp,
  useYBar: booleanProp,
  barFade: Number,
  barClass: classProp,
  autoplay: booleanNumberProp,
  playWaiting: Number,
  onBeforeScroll: Function,
  appear: booleanProp,
  barDuration: Number,
  useBarTrack: booleanProp,
  wrapperTag: String,
  onResize: eventProp(),
  onXEnabledChange: eventProp(),
  onYEnabledChange: eventProp(),
  onWheel: eventProp(),
  onScrollStart: eventProp(),
  onScroll: eventProp(),
  onScrollEnd: eventProp(),
  onBarScrollStart: eventProp(),
  onBarScroll: eventProp(),
  onBarScrollEnd: eventProp()
});
export { nativeScrollProps };
