// vexip-ui/es/components/form/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanProp, sizeProp, eventProp } from "../../common/config/src/props.mjs";
const formProps = buildProps({
  method: String,
  action: String,
  model: Object,
  rules: Object,
  labelWidth: [Number, String],
  labelAlign: String,
  allRequired: booleanProp,
  labelSuffix: String,
  hideAsterisk: booleanProp,
  validateAll: booleanProp,
  hideLabel: booleanProp,
  disabled: booleanProp,
  loading: booleanProp,
  size: sizeProp,
  inline: booleanProp,
  gap: [Number, Array],
  justify: String,
  align: String
});
const mediaProp = [Number, Object];
const formItemProps = buildProps({
  label: String,
  prop: String,
  rules: [Object, Array],
  labelWidth: Number,
  required: booleanProp,
  htmlFor: String,
  errorTransition: String,
  defaultValue: Object,
  hideErrorTip: booleanProp,
  validateAll: booleanProp,
  hideAsterisk: booleanProp,
  hideLabel: booleanProp,
  action: booleanProp,
  help: String,
  pure: booleanProp,
  span: Number,
  offset: Number,
  push: Number,
  pull: Number,
  order: Number,
  xs: mediaProp,
  sm: mediaProp,
  md: mediaProp,
  lg: mediaProp,
  xl: mediaProp,
  xxl: mediaProp,
  flex: [Number, String]
});
const formSubmitProps = buildProps({
  size: sizeProp,
  type: String,
  label: String,
  dashed: booleanProp,
  text: booleanProp,
  simple: booleanProp,
  ghost: booleanProp,
  disabled: booleanProp,
  circle: booleanProp,
  loadingIcon: Object,
  loadingSpin: booleanProp,
  icon: Object,
  color: String,
  buttonType: String,
  block: booleanProp,
  tag: String,
  onBeforeSubmit: Function,
  onSubmit: eventProp(),
  onError: eventProp()
});
const formResetProps = buildProps({
  size: sizeProp,
  type: String,
  label: String,
  dashed: booleanProp,
  text: booleanProp,
  simple: booleanProp,
  ghost: booleanProp,
  disabled: booleanProp,
  loading: booleanProp,
  circle: booleanProp,
  loadingIcon: Object,
  loadingSpin: booleanProp,
  icon: Object,
  color: String,
  buttonType: String,
  block: booleanProp,
  tag: String,
  onBeforeReset: Function,
  onReset: eventProp()
});
export { formItemProps, formProps, formResetProps, formSubmitProps };
