// vexip-ui/es/components/tree/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanProp, eventProp } from "../../common/config/src/props.mjs";
const treeProps = buildProps({
  arrow: {
    type: [Boolean, String],
    default: null
  },
  data: Array,
  noBuildTree: booleanProp,
  emptyTip: String,
  disabled: booleanProp,
  readonly: booleanProp,
  checkbox: booleanProp,
  suffixCheckbox: booleanProp,
  renderer: Function,
  multiple: booleanProp,
  indent: [String, Number],
  accordion: booleanProp,
  draggable: booleanProp,
  appear: booleanProp,
  floorSelect: booleanProp,
  onAsyncLoad: Function,
  cacheNode: booleanProp,
  rootId: [String, Number],
  keyConfig: Object,
  noCascaded: booleanProp,
  filter: [String, Function],
  ignoreCase: booleanProp,
  nodeProps: [Object, Function],
  onNodeChange: eventProp(),
  onNodeClick: eventProp(),
  onNodeSelect: eventProp(),
  onNodeCancel: eventProp(),
  onNodeExpand: eventProp(),
  onNodeReduce: eventProp(),
  onDragStart: eventProp(),
  onDragOver: eventProp(),
  onDrop: eventProp(),
  onDragEnd: eventProp(),
  onLabelClick: eventProp()
});
export { treeProps };
