// vexip-ui/es/components/collapse/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, booleanProp, eventProp, styleProp } from "../../common/config/src/props.mjs";
const collapseProps = buildProps({
  expanded: [String, Number, Array],
  card: booleanProp,
  accordion: booleanProp,
  arrowType: String,
  ghost: booleanProp,
  onChange: eventProp()
});
const collapsePanelProps = buildProps({
  label: [String, Number],
  title: String,
  disabled: booleanProp,
  contentStyle: styleProp,
  expanded: booleanProp,
  card: booleanProp,
  arrowType: String,
  icon: Object,
  ghost: booleanProp,
  onToggle: eventProp()
});
const collapseTransitionProps = buildProps({
  appear: booleanProp,
  mode: String,
  horizontal: booleanProp,
  duration: Number,
  timing: String,
  fadeEffect: booleanProp,
  onBeforeEnter: eventProp(),
  onEnter: eventProp(),
  onAfterEnter: eventProp(),
  onEnterCancelled: eventProp(),
  onBeforeLeave: eventProp(),
  onLeave: eventProp(),
  onAfterLeave: eventProp(),
  onLeaveCancelled: eventProp()
});
export { collapsePanelProps, collapseProps, collapseTransitionProps };
