// vexip-ui/es/components/input/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanProp, classProp, eventProp } from "../../common/config/src/props.mjs";
const inputProps = buildProps({
  size: sizeProp,
  state: stateProp,
  type: String,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  formatter: Function,
  value: [String, Number],
  placeholder: String,
  autofocus: booleanProp,
  spellcheck: booleanProp,
  autocomplete: booleanProp,
  readonly: booleanProp,
  disabled: booleanProp,
  inputClass: classProp,
  debounce: booleanProp,
  maxLength: Number,
  before: String,
  after: String,
  plainPassword: booleanProp,
  clearable: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  transparent: booleanProp,
  sync: booleanProp,
  onFocus: eventProp(),
  onBlur: eventProp(),
  onInput: eventProp(),
  onChange: eventProp(),
  onEnter: eventProp(),
  onClear: eventProp(),
  onPrefixClick: eventProp(),
  onSuffixClick: eventProp(),
  onKeyDown: eventProp(),
  onKeyPress: eventProp(),
  onKeyUp: eventProp()
});
export { inputProps };
