// vexip-ui/es/components/color-picker/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanProp, booleanStringProp, eventProp } from "../../common/config/src/props.mjs";
const colorPickerProps = buildProps({
  size: sizeProp,
  state: stateProp,
  value: [String, Object],
  visible: booleanProp,
  format: String,
  alpha: booleanProp,
  disabled: booleanProp,
  transitionName: String,
  noInput: booleanProp,
  shortcut: {
    type: [Boolean, Array],
    default: null
  },
  placement: String,
  transfer: booleanStringProp,
  outsideClose: booleanProp,
  clearable: booleanProp,
  cancelText: String,
  confirmText: String,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  noSuffix: booleanProp,
  staticSuffix: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  onToggle: eventProp(),
  onClickOutside: eventProp(),
  onOutsideClose: eventProp(),
  onClear: eventProp(),
  onChange: eventProp(),
  onShortcut: eventProp()
});
export { colorPickerProps };
