// vexip-ui/es/components/auto-complete/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanStringProp, booleanProp, eventProp } from "../../common/config/src/props.mjs";
const autoCompleteProps = buildProps({
  size: sizeProp,
  state: stateProp,
  transfer: booleanStringProp,
  value: [String, Number],
  options: Array,
  filter: {
    type: [Boolean, Function],
    default: null
  },
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  placeholder: String,
  disabled: booleanProp,
  transitionName: String,
  dropDisabled: booleanProp,
  placement: String,
  clearable: booleanProp,
  ignoreCase: booleanProp,
  autofocus: booleanProp,
  spellcheck: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  transparent: booleanProp,
  keyConfig: Object,
  onSelect: eventProp(),
  onInput: eventProp(),
  onChange: eventProp(),
  onToggle: eventProp(),
  onEnter: eventProp(),
  onClear: eventProp()
});
export { autoCompleteProps };
