// vexip-ui/es/components/number-input/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, sizeProp, stateProp, booleanProp, classProp, eventProp } from "../../common/config/src/props.mjs";
const numberInputProps = buildProps({
  size: sizeProp,
  state: stateProp,
  prefix: Object,
  prefixColor: String,
  suffix: Object,
  suffixColor: String,
  formatter: Function,
  value: Number,
  min: Number,
  max: Number,
  placeholder: String,
  autofocus: booleanProp,
  spellcheck: booleanProp,
  autocomplete: booleanProp,
  precision: Number,
  readonly: booleanProp,
  step: Number,
  ctrlStep: Number,
  shiftStep: Number,
  altStep: Number,
  disabled: booleanProp,
  inputClass: classProp,
  debounce: booleanProp,
  clearable: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  sync: booleanProp,
  controlType: String,
  onFocus: eventProp(),
  onBlur: eventProp(),
  onInput: eventProp(),
  onChange: eventProp(),
  onEnter: eventProp(),
  onClear: eventProp(),
  onPrefixClick: eventProp(),
  onSuffixClick: eventProp(),
  onKeyDown: eventProp(),
  onKeyPress: eventProp(),
  onKeyUp: eventProp()
});
export { numberInputProps };
