// vexip-ui/es/components/upload/props.mjs
import "../../common/config/src/index.mjs";
import { buildProps, stateProp, booleanProp, styleProp, eventProp } from "../../common/config/src/props.mjs";
const uploadProps = buildProps({
  state: stateProp,
  url: String,
  fileList: Array,
  multiple: booleanProp,
  tip: String,
  accept: [String, Array],
  filter: [String, Array],
  maxSize: Number,
  field: String,
  data: Object,
  headers: Object,
  withCredentials: booleanProp,
  manual: booleanProp,
  hiddenFiles: booleanProp,
  countLimit: Number,
  allowDrag: booleanProp,
  onBeforeUpload: Function,
  onBeforeSelect: Function,
  iconRenderer: Function,
  selectToAdd: booleanProp,
  listType: String,
  block: booleanProp,
  loadingText: String,
  directory: booleanProp,
  pathField: String,
  disabledClick: booleanProp,
  buttonLabel: String,
  disabled: booleanProp,
  loading: booleanProp,
  loadingIcon: Object,
  loadingLock: booleanProp,
  loadingSpin: booleanProp,
  image: booleanProp,
  defaultFiles: Array,
  canPreview: Function,
  listStyle: styleProp,
  onExceed: eventProp(),
  onChange: eventProp(),
  onFilterError: eventProp(),
  onSizeError: eventProp(),
  onDelete: eventProp(),
  onPreview: eventProp(),
  onProgress: eventProp(),
  onSuccess: eventProp(),
  onError: eventProp()
});
const uploadListProps = buildProps({
  files: Array,
  selectToAdd: booleanProp,
  iconRenderer: Function,
  type: String,
  loadingText: String,
  style: styleProp,
  precision: Number,
  canPreview: Function,
  onDelete: eventProp(),
  onPreview: eventProp()
});
const uploadFileProps = buildProps({
  file: Object,
  iconRenderer: Function,
  listType: String,
  loadingText: String,
  selectToAdd: booleanProp,
  precision: Number,
  canPreview: Function,
  onDelete: eventProp(),
  onPreview: eventProp()
});
export { uploadFileProps, uploadListProps, uploadProps };
